packages/flutter/lib/src/widgets/actions.dart

```diff
@@ -0,0 +1,308 @@
+///    up key combinations in order to invoke actions.
+///  - [Actions], which is a widget that defines a map of [Intent] to [Action]
+///    and allows redefining of actions for its descendants.
+///  - [ActionDispatcher], a class that takes an [Action] and invokes it using a
+///    [FocusNode] for context.
+abstract class Action extends Diagnosticable {
+  /// A const constructor for an [Action].
+  ///
+  /// The [intentKey] parameter must not be null.
+  const Action(this.intentKey) : assert(intentKey != null);
+
+  /// The unique key for this action.
+  ///
+  /// This key will be used to map to this action in an [ActionDispatcher].
+  final LocalKey intentKey;
+
+  /// Called when the action is to be performed.
+  ///
+  /// This is called by the [ActionDispatcher] when an action is accepted by a
+  /// [FocusNode] by returning true from its `onAction` callback, or when an
+  /// action is invoked using [ActionDispatcher.invokeAction].
+  ///
+  /// This method is only meant to be invoked by an [ActionDispatcher], or by
+  /// subclasses.
+  ///
+  /// Actions invoked directly with [ActionDispatcher.invokeAction] may receive a
+  /// null `node`. If the information available from a focus node is
+  /// needed in the action, use [ActionDispatcher.invokeFocusedAction] instead.
+  @protected
+  @mustCallSuper
+  void invoke(FocusNode node, covariant Intent tag);
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<LocalKey>('intentKey', intentKey));
+  }
+}
+
+/// The signature of a callback accepted by [CallbackAction].
+typedef OnInvokeCallback = void Function(FocusNode node, Intent tag);
+
+/// An [Action] that takes a callback in order to configure it without having to
+/// subclass it.
+///
+/// See also:
+///
+///  - [Shortcuts], which is a widget that contains a key map, in which it looks
+///    up key combinations in order to invoke actions.
+///  - [Actions], which is a widget that defines a map of [Intent] to [Action]
+///    and allows redefining of actions for its descendants.
+///  - [ActionDispatcher], a class that takes an [Action] and invokes it using a
+///    [FocusNode] for context.
+class CallbackAction extends Action {
+  /// A const constructor for an [Action].
+  ///
+  /// The `intentKey` and [onInvoke] parameters must not be null.
+  /// The [onInvoke] parameter is required.
+  const CallbackAction(LocalKey intentKey, {@required this.onInvoke})
+      : assert(onInvoke != null),
+        super(intentKey);
+
+  /// The callback to be called when invoked.
+  ///
+  /// Must not be null.
+  @protected
+  final OnInvokeCallback onInvoke;
+
+  @override
+  void invoke(FocusNode node, Intent tag) => onInvoke.call(node, tag);
+}
+
+/// An action manager that simply invokes the actions given to it.
+class ActionDispatcher extends Diagnosticable {
```

Comment:
What is the purpose of having this as a class? It doesn't look like it is dispatching so much as just a small wrapper around `action.invoke`.  Could this just a be function? Or are you anticipating subclasses of `ActionDispatcher`?

Reply:
Yes, I'm anticipating having subclasses. Take a look at `dev/manual_tests/lib/actions.dart` and you'll see an example of this.  Having the ability to subclass and replace your own `ActionDispatcher` means you can have complete control over the dispatch.  I could have this be just a single "invoke" function that you supply, but it feels more like a class because there could be other operations we want to add in the future (mayby `canInvoke`, or something), and this gives us a place to put it.